service-config/src/main/java/com/palantir/config/service/ServiceConfiguration.java

```diff
@@ -43,4 +43,23 @@
      * A list of service URIs.
      */
     public abstract List<String> uris();
+
+    // hides implementation details
+    public static Builder builder() {
+        return ImmutableServiceConfiguration.builder();
+    }
+
+    // hides implementation details
+    public interface Builder {
```

Comment:
Just played around with this more locally, and can confirm what @uschi2000 and @pnepywoda said -- basically, the `extends Immutable*.Builder {}` approach should work fine. @qinfchen , what was the scenario where this didn't work? I tried using the approach described and consuming from an external project without Immutables and everything still compiled/worked.

Philosophically, the re-defining approach is cleaner because it does hide the impl details. For example, if you change the templating so that the generated immutable classes have a different naming scheme, the `extends Immutable*.Builder {}` approach will break in source, while the former approach will not. Also, if there's a situation where the source is published for the interface/abstract class but not for the generated class, then defining the interface in the declaring class makes more sense.

Ultimately it seems like either approach works fine in almost all cases, so I think this largely comes down to a matter of taste.